FancyBatching.md

```diff
@@ -0,0 +1,505 @@
+        - `event Action<IndexDocumentsAction<T>, IndexingResult, Exception, CancellationToken> ActionFailed;`
+        - `event Func<IndexDocumentsAction<T>, IndexingResult, Exception, CancellationToken, Task> ActionFailedAsync;`
+
+            Either the `IndexingResult` or `Exception` may be null for failures
+            depending on the type of failure and how it was surfaced.
+
+            We'll raise either ActionCompleted or ActionFailed events.  You'd
+            need to handle both to implement checkpointing.
+
+            We will not throw any exceptions from the pipeline.  If a failure
+            is triggered from a timer, there's no good way for a customer to
+            catch it.  Everything will go through the notification events
+            instead.  If we find the need for request level failures, we could
+            add a `public event Action<Exception, CancellationToken> BatchFailed;`
+            or similar.
+
+            If a sync method like `Flush` is called, we'll raise the sync
+            `ActionSent` event.  If an aysnc method like `FlushAsync` is called,
+            we'll raise `ActionSentAsync`.  In the event of a timer elapsing,
+            we will always raise the async event (note that we have a separate
+            `CancellationToken` for this scenario too.)
+
+            Because we can't throw once we've entered the publisher, we'll
+            validate that the customer is handling the correct events before
+            leaving their call stack.  This should be done from every "Add
+            Action" method and `Flush` because handlers can be added and removed.
+
+            - If I'm calling an async method, we need to make sure every async
+            event has handlers if the corresponding sync event also has handlers.
+
+            - If I'm calling a sync method, we need to make sure every sync
+            event has handlers if the corresponding async event has handlers.
+
+            - If I'm calling a sync method when `AutoFlush == true` and
+            `AutoFlushInterval != null`, then I need to have parity between the
+            handlers for both sync and async events since any immediate
+            publishing will happen with a sync handler but an elapsed timer
+            publish will happen with an async handler.
+
+            When raising an async event, we'll collect all the returned tasks
+            from each handler and `Task.WaitAll` to ensure they've all run to
+            completion.
+
+    - Cleaning up with `Dispose`
+
+        - `void Dispose()` will call the publisher's `Dispose`, which does a
+        blocking flush.
+
+        - `Task DisposeAsync()` will call the publisher's `DisposeAsync`, which
+        does a blocking flush.
+
+        - We need the publisher to block on these particular calls because the
+        application might exit as soon as `Dispose` returns.
+
+        - We'll set `bool _disposed = true` and throw `ObjectDisposedException`
```

Comment:
When `_disposed == true` does calling `Dispose` again trigger this exception as well? I know you mention attempting to call any method will raise it but what happens when multiple threads attempt disposing at once.

Reply:
No.  Per https://docs.microsoft.com/en-us/visualstudio/code-quality/ca2202?view=vs-2019#rule-description:

> **A correctly implemented Dispose method can be called multiple times without throwing an exception.** However, this is not guaranteed and to avoid generating a System.ObjectDisposedException you should not call Dispose more than one time on an object.

When multiple threads call `Dispose` at the same time, the first one will take the monitor and the others will block.  Once the first one is done, the others will be let in one at a time so it kind of ends up being the double lock check pattern.